shell/server/proxy.js

```diff
@@ -310,14 +319,25 @@ function openSessionInternal(grainId, user, title, apiToken) {
 
   updateLastActive(grainId, userId);
 
-  var isOwner = grainInfo.owner === userId;
+  var sessionId = generateSessionId(grainId, userId, cachedSalt);
+  var session = Sessions.findOne({_id: sessionId});
+  if (session) {
+    // TODO(someday): also do some more checks for anonymous sessions (sessions without a userId).
+    if ((session.userId && session.userId !== userId) ||
+        (session.grainId !== grainId)) {
+      cachedSalt = Random.id(22);
```

Comment:
Also, is this the only place where a new cachedSalt is actually generated? Shouldn't a new cachedSalt always be generated if the client passed `undefined`?